### PR TITLE
Edit worker page

### PR DESCRIPTION
**Pull request by opensourcerules**
Added edit worker page/functionality

### CODE REVIEW

File: app/controllers/WorkerController.php [hunk 1]
```diff
@@ -81,4 +81,27 @@ public function registerAction(): void
 
         $this->view->errors = $errors;
     }
+
+    /**
+     * If admin is logged in, updates worker via worker service.
+     */
+    public function editAction(): void
+    {
+        if (false === $this->authenticationService->isLoggedIn()) {
```
Comment (andreiisoc) [inline app/controllers/WorkerController.php]: authentication seems to be on every controller and repeating. this seems too look like a middleware

File: app/services/Worker.php [hunk 2]
```diff
@@ -48,4 +48,39 @@ public function register(array $worker): object
...
         return $errors;
     }
+
+    /**
+     *Gets all the workers from db via model
+     * @return \Phalcon\Mvc\Model\ResultsetInterface;
+     */
+    public function getWorkers(): object
+    {
+        return $this->workerModel->find([
```
Comment (andreiisoc) [inline app/services/Worker.php]: avoid mysql concat. create a new method in model that does this for you

Comment (andreiisoc) [reply app/services/Worker.php]: also first name last name is not allInfo... is just: full name :)

File: app/services/Worker.php [hunk 3]
```diff
@@ -48,4 +48,39 @@ public function register(array $worker): object
...
+            'columns' => "id, CONCAT(firstName, ' ', lastName, ' : ', email) AS allInfo",
+            'order' => 'allInfo',
+        ]);
+    }
+
+    /**
+     * Updates the worker via model in db
```
Comment (andreiisoc) [inline app/services/Worker.php]: avoid shorting in comments: db = databse

File: app/middleware/Redirect.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+use Phalcon\Application;
+use Phalcon\Events\Event;
+use GanttDashboard\App\Services\Authentication as AuthenticationService;
+
+/**
+ * Redirect
+ *
+ * Checks the request and redirects the user somewhere else if needed
+ */
+class Redirect
```
Comment (andreiisoc) [inline app/middleware/Redirect.php]: i believe that this is an authentication middleware that checks if a user is authenticated and if not it gets redirected

File: app/middleware/Redirect.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+namespace GanttDashboard\App\Middleware;
+
+use Phalcon\Application;
+use Phalcon\Events\Event;
+use GanttDashboard\App\Services\Authentication as AuthenticationService;
+
+/**
+ * Redirect
+ *
+ * Checks the request and redirects the user somewhere else if needed
```
Comment (andreiisoc) [inline app/middleware/Redirect.php]: this does not check the request. it checks if a user is authenticated

File: app/middleware/Redirect.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+     * Before anything happens
+     *
+     * @param Event $event
+     * @param Application $application
+     *
+     * @return bool
+     */
+    public function beforeSendResponse(Event $event, Application $application)
+    {
+        $restrictedUri = [
```
Comment (andreiisoc) [inline app/middleware/Redirect.php]: are you really going to put here all restricted urls?
also try to check ACL and see if that integration with a middleware fits better

File: app/models/Workers.php [hunk 7]
```diff
@@ -115,6 +115,22 @@ public function getEmail(): string
         return $this->email;
     }
 
+    /**
+     * Returns multidimensional array with id and full name of all workers
+     * @return array
+     */
+    public function getWorkersWithFullName()
```
Comment (andreiisoc) [inline app/models/Workers.php]: do you have workers without full name? I believe not....
This method seems to me that is get all workers and it orders the by first name, then last name. Name you methods accordingly, and also the php doc.

File: app/models/Workers.php [hunk 8]
```diff
@@ -115,6 +115,22 @@ public function getEmail(): string
...
+    {
+        $fullNameWorkers = [];
+        $allWorkers = Workers::find(['order' => 'firstName, lastName'])->toArray();
```
Comment (andreiisoc) [inline app/models/Workers.php]: avoid ->toArray()
If it OOP, then don't use array... use objects instead

File: app/models/Workers.php [hunk 9]
```diff
@@ -115,6 +115,22 @@ public function getEmail(): string
...
+
+        foreach ($allWorkers as $worker) {
```
Comment (andreiisoc) [inline app/models/Workers.php]: you should get your query full serve and never needed to loop over it.
Looping over results should be made by views

Comment (andreiisoc) [reply app/models/Workers.php]: also remember that model does the queries.... manipulating it is part of views/service
in this case this should never be made... the data should be returned as it is to the views

File: app/views/worker/edit.phtml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        </div>
+    </div>
+
+    <div class='form-group'>
+        <label for='lastName'>Last name</label>
+        <div class='row justify-content-center'>
+            <div class='col-xs-12 col-sm-6 col-md-4 col-lg-3'>
+                {% if errors is defined and errors.filter('lastName')|length > 0 %}
+                <div class='text-danger'>
+                    {{ errors.filter('lastName')[0].getMessage() }}
```
Comment (andreiisoc) [inline app/views/worker/edit.phtml]: you should never access array by keys manually
eg: $array[0]... loop over them or find other solutions

File: app/views/worker/edit.phtml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        </div>
+    </div>
+
+    <div class='form-group'>
+        <label for='firstName'>First name</label>
+        <div class='row justify-content-center'>
+            <div class='col-xs-12 col-sm-6 col-md-4 col-lg-3'>
+                {% if errors is defined and  errors.filter('firstName')|length > 0 %}
+                <div class='text-danger'>
+                    {{ errors.filter('firstName')[0].getMessage() }}
```
Comment (andreiisoc) [inline app/views/worker/edit.phtml]: same as above

File: app/views/worker/edit.phtml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        </div>
+    </div>
+
+    <div class='form-group'>
+        <label for='email'>Email</label>
+        <div class='row justify-content-center'>
+            <div class='col-xs-12 col-sm-6 col-md-4 col-lg-3'>
+                {% if errors is defined and  errors.filter('email')|length > 0 %}
+                <div class='text-danger'>
+                    {{ errors.filter('email')[0].getMessage() }}
```
Comment (andreiisoc) [inline app/views/worker/edit.phtml]: same as above

File: public/index.php [hunk 13]
```diff
@@ -63,10 +63,16 @@
      */
     include APP_PATH . '/config/loader.php';
 
+    /**
+     * Include Middleware
+     */
+    include APP_PATH . '/config/middleware.php';
```
Comment (andreiisoc) [inline public/index.php]: the implementation of the middleware does not seems right. please stop to use bad examples from stack overlow.
you should never need to include files. Phalcon uses composer and PSR4 standard, and it should not required includes.
Once again.... stack overflow sucks... next copy/past = block stackoverlow

**Review by andreiisoc — COMMENTED** [review]
please stop using stack overflow

File: app/middleware/Redirect.php (hunk 6 above)
Comment (flaviu-chelaru) [reply app/middleware/Redirect.php]: Replace the event with something else: you need to do all the checks BEFORE you are handling the request in the contrller, not after

File: app/config/router.php [hunk 14]
```diff
@@ -1,45 +1,66 @@
...
+$index = new RouterGroup([
     'controller' => 'Index',
-    'action'     => 'notFound',
 ]);
 
+$index->add('/', [
+    'action' => 'index',
+])->setName('home');
+
+$index->add('/index/notfound', [
```
Comment (flaviu-chelaru) [inline app/config/router.php]: use addGet or addPost. If you have multiple VERBS for a route, use add, but use via() as well

File: app/controllers/IndexController.php [hunk 15]
```diff
@@ -6,8 +6,12 @@
 
 class IndexController extends Controller
 {
+    /**
+     * Define the index action
+     */
     public function indexAction(): void
     {
+        $this->flashSession->warning('Under construction');
```
Comment (flaviu-chelaru) [inline app/controllers/IndexController.php]: controllers should return ResponseInterface implementations.

File: app/middleware/Authentication.php [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        $this->router = $router;
+    }
+
+    /**
+     * Checks if visitor has permissions to view the requested url
+     * @param Event $event
+     * @param Application $application
+     * @return bool
+     */
+    public function beforeSendResponse(Event $event, Application $application): bool
```
Comment (flaviu-chelaru) [inline app/middleware/Authentication.php]: wrong event - use something that triggers before you run the controller action

File: app/providers/MyDi.php [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+        $this->register(new RegisterEventsManager());
+        $this->register(new RegisterAclList());
+        $this->register(new RegisterSession());
+        $this->register(new RegisterServices());
+        $this->register(new RegisterMiddlewares());
+        $this->register(new RegisterFlash());
+        $this->register(new RegisterUrl());
+        $this->register(new RegisterDb());
+        $this->register(new RegisterValidators());
+        $this->setShared('request', Request::class);
```
Comment (flaviu-chelaru) [inline app/providers/MyDi.php]: pick one: service provider or register directly in the di. Don't use both

File: app/providers/RegisterDb.php [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+class RegisterDb implements ServiceProviderInterface
+{
+    /**
+     * Database connection is created based on the parameters defined in the configuration file
+     * @param DiInterface $di
+     * @return void
+     */
+    public function register(DiInterface $di): void
+    {
+        $di->setShared('db', function () use ($di) {
```
Comment (flaviu-chelaru) [inline app/providers/RegisterDb.php:18]: use Phalcon\Db\Adapter\Pdo\Factory;

File: app/providers/RegisterRouter.php [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * Registers router
+     * @param DiInterface $di\
+     * @return void
+     */
+    public function register(DiInterface $di): void
+    {
+        $di->setShared('router', function () {
+            return include APP_PATH . '/config/router.php';
```
Comment (flaviu-chelaru) [inline app/providers/RegisterRouter.php]: don't use include / require

File: app/providers/RegisterUrl.php [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+{
+    /**
+     * Register the URL component that is used to generate all kind of urls in the application
+     * @param DiInterface $di
+     * @return void
+     */
+    public function register(DiInterface $di): void
+    {
+        $di->setShared('url', function () use ($di) {
+            $config = $di->get('config');
```
Comment (flaviu-chelaru) [inline app/providers/RegisterUrl.php:19]: works with $this->get('config') - no need to add the $di in USE

File: app/services/Worker.php [hunk 21]
```diff
@@ -13,49 +13,59 @@ class Worker
...
+        ]);
+    }
+
+    /**
+     * Gets worker from database via model
+     * @param int $id
+     * @return \Phalcon\MVC\Model\ResultsetInterface
      */
-    public function getWorkers(): array
+    public function getWorker(int $id): object
```
Comment (flaviu-chelaru) [inline app/services/Worker.php]: "object" is too general return type. Use specific or interface where available

File: public/index.php [hunk 22]
```diff
@@ -36,45 +36,31 @@
...
     /**
-     * The FactoryDefault Dependency Injector automatically registers
-     * the services that provide a full stack framework.
+     * Use composer autoloader to load vendor classes
      */
-    $di = new FactoryDefault();
+    require_once BASE_PATH . "/vendor/autoload.php";
+
+    $di = new MyDi();
+    $di->initialize();
```
Comment (flaviu-chelaru) [inline public/index.php]: use __construct and call the parent::__construct()

**Review by flaviu-chelaru — CHANGES_REQUESTED** [review]
Fix the requested stuff before merging

File: app/providers/RegisterConfig.php [hunk 23]
```diff
@@ -15,7 +16,9 @@ class RegisterConfig implements ServiceProviderInterface
     public function register(DiInterface $di): void
     {
         $di->setShared('config', function () {
-            return include APP_PATH . '/config/config.php';
+            require APP_PATH . '/config/config.php';
+
+            return new Config($settings);
```
Comment (flaviu-chelaru) [inline app/providers/RegisterConfig.php]: use Phalcon\Config\Factory; to load data from a config file.

File: app/services/Worker.php [hunk 24]
```diff
@@ -4,6 +4,9 @@
 
 use GanttDashboard\App\Models\Workers;
 use GanttDashboard\App\Validators\Worker as WorkerValidator;
+use Phalcon\MVC\Model\ResultsetInterface;
+use Phalcon\Validation\Message\Group as MessageGroup;
+use \Phalcon\Mvc\Model;
```
Comment (flaviu-chelaru) [inline app/services/Worker.php]: remove the leading backslash

File: app/services/Worker.php [hunk 25]
```diff
@@ -71,13 +74,17 @@ public function getWorker(int $id): object
...
      */
-    public function edit(array $workerUpdate): object
+    public function edit(array $workerUpdate): MessageGroup
     {
         $errors = $this->workerValidator->validate($workerUpdate);
 
         if (0 == $errors->count()) {
+
+            /**
+             * @var $worker Workers
```
Comment (flaviu-chelaru) [inline app/services/Worker.php]: this shouldn't be on multiple lines - put inline

File: app/views/worker/beforeEdit.phtml [hunk 26]
```diff
@@ -15,15 +15,15 @@
     </div>
 
 {% for worker in workers%}
-        <div class="row">
+        <div class="row border border-top-0 p-1">
             <div class="col">
-                {{ link_to('worker/edit/' ~  worker.id, worker.firstName) }}
+                {{ link_to('worker/edit/' ~  worker.id, worker.firstName, 'class': 'text-dark') }}
```
Comment (flaviu-chelaru) [inline app/views/worker/beforeEdit.phtml]: replace with the name routes
From documentation:

`{{ link_to(['for': 'show-product', 'id': 123, 'name': 'carrots'], 'Show') }}
`

**Review by flaviu-chelaru — CHANGES_REQUESTED** [review]
Please make the requested changes when possible. Thanks

**Review by flaviu-chelaru — APPROVED** [review]

**Pull request merged**